### DISCUSSION TITLE
Transcend Vault

### DISCUSSION THREAD

**Discussion by envoy1084** [NillionNetwork/discussions#62] [Show and tell]
### Show and tell project type

Builder Bounty Submission

### Github Repo Link

https://github.com/Envoy-VC/transcend-nillion

### Video Walkthough Link

https://www.youtube.com/watch?v=m1_AWJ8iMpY

### Project Description

Transcend Vault is a Multi-Peer Secrets Vault which is gated by a secure Multi-Party Computation (MPC) based Biometric Authentication System.

It uses Nillion for Blind Computation of matching and finding Euclidean Similarity between Face Descriptors.

Peer to Peer Network is established using libp2p and OrbitDB for storing the data over a P2P Database.

### Secure Vault Storage

The projects acts as a secure multi-peer vault, so it can be used by organizations for secrets storage and retrieval in workflows. Multi-peer is possible using `Libp2p` and `OrbitDB` which help in P2P Communication and Database Syncing.

### Multi-party Biometric Authentication

The entire Vault is gated by a secure MPC based Biometric Authentication carried out by computing similarity between Face Descriptors blindly.

### How does the project use Nillion? Describe and link to any Nada programs

- Nillion Programs are used to compute `euclidean_distance` between two sets of face descriptors.
- Nillion is used as secret storage and retrieval for vault.

Here is the basic Logic for the program that calculates `euclidean_distance`

```python
def sqrt(num: SecretInteger):
    guess = num
    for _ in range(12):
        guess = (guess + num / guess) / Integer(2)
    return guess


def get_euclidean_distance(actual: list[list[SecretInteger]], given: list[SecretInteger]) -> Integer:
    distances: list[Integer] = []
    for encoding in actual:
        distance = Integer(0)
        for i in range(DESCRIPTOR_SIZE):
            diff = encoding[i] - given[i]
            squared_diff = diff * diff
            distance += squared_diff
        distances.append(sqrt(distance))
    return distances[0]
```

Also `sqrt` is calculated using Iterative Newton's Method for Square Root till 12 Iterations.

When we calculate Face Descriptor we get a `128` size `FloatArray`, but as floats are not supported in nada we multiply each value by 1000 and round it off. so now all values are in range `(0,1000]`. 

Entire Nada Program can be found at: https://github.com/Envoy-VC/transcend-nillion/blob/main/packages/programs/src/main.py

### Is there anything else you want to share?

For `sqrt` function if I increase the iterations to above 14, it takes long time to build the program, not sure why?

### Optional - Link your project and team members' social handles

https://twitter.com/Envoy_1084

### Optional - Team ETH Address(es)

0xBF4979305B43B0eB5Bb6a5C67ffB89408803d3e1

**Comment by oceans404** [comment]
Congratulations @Envoy-VC!!! You won one of our tinybio bounties. Can you confirm that the ETH address you provided is correct so we can send your prize: 0xBF4979305B43B0eB5Bb6a5C67ffB89408803d3e1

tinydemo winners announcement tweet: https://x.com/nillionnetwork/status/1820606184895947087

**Reply by envoy1084** [reply]
Thanks, yes that is the correct address.

**Reply by Nillion-Engineering** [reply]
Hey, Claire from the Nillion Association here! Thanks for building on Nillion. Here's the link to your prize. Well done!

https://etherscan.io/tx/0x479502499800bbe9d6c936dc9b05aaf23819227bf2283ada10dcff8184e3d1ac
